GoogleForCreators/web-stories-wp
# Dashboard: Update dashboard grid space according to wordpress menu width

# Summary
Make sure dashboard is considering the total space available to stories regardless of wordpress left nav being collapsed or expanded 

## Relevant Technical Choices (originally)
- Initially I was only checking for the collapsed wordpress left nav, but when the menu's expanded it shrinks the dashboard's total available space. Since this is outside of our jurisdiction with the dashboard app, need to control this by watching for clicks on the menu and then recalculate the space available for the pageSize getting used in grid 
- Setting wordpress nav element `#adminmenu` as a ref, setting .current immediately to the element because we don't need it to cause any extra renders. 
- New `useLayoutEffect` to have an eventListener on nav for clicks. Debounce clicks and set width off of that instance of current.offsetWidth for nav.
- New `useLayoutEffect` to update the available container space with new wpMenuWidth when it changes. 
- wpMenuWidth defaults to 0

## Updated PR 
- this escalated from a 2 file change pr to 22 file changes - here is why: 
- Pascal made the request that this util try not to rely on wordpress and made me aware of a custom hook in the editStory directory called `useResizeEffect` that he'd rather get used. Originally, this usePagePreviewSize util was just a small thing to get the story sizes so we could render them - it turned out we needed that size for the grid not just the stories and need that util to be watching for the size available to get the proper size in pixels of stories in a grid setting as well as isolated. Originally doing this with just the one watcher for the container size was enough, but that didn't take into consideration the dashboard available width when the wordpress menus were expanded. 
- Implementing the `useResizeEffect` hook drastically reduced the checks we need to handle to get the pageSize and simplified the code so I think this update was worthwhile 
- While refactoring this code, I figured it was a good time to also clean up how many times the util is getting used and passing the pageSize into the grid elements that need it (namely the preview since it needs pixel size to overlay properly and prevent pageSize getting called repeatedly - this is where all of the extra file diff comes from - I had to update storybook and tests tied to this structure. 
- It's also important to note that we still have to tie our resizeEffect to a ref based on `wpbody` because this id exists outside of just the dashboard it can tell when the wordpress part of the page as a whole is also resized (meaning viewport AS WELL as menus expanding or collapsing). I spent some time yesterday experimenting with IDs inside of the dashboard app to keep it contained and it just doesn't work. This is something that will have to change when the dashboard is pulled out of wordpress, but it's 1 line to change and there's a constant in `pageStructure` making it pretty clear. 


## User-facing changes
- Grid items should now fill the total grid space regardless of if wordpress menu is expanded or collapsed

![wp menu toggle resize](https://user-images.githubusercontent.com/10720454/82697844-8efb3080-9c1e-11ea-8552-2ec7b5671b00.gif)


## Testing Instructions
- Toggle the wordpress menu open and closed and see that the grid updates the item size to fill the available space. 
- See that [storybook/myStories/content](http://localhost:59746/?path=/story/dashboard-components-mystories-content--default) works still (it has no wordpress menu, so this is a good spot to see that 0 come into play)
- see that list and grid views are as expected 



---



Fixes #
